src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -100,75 +100,50 @@ Task OnUnconfiguredProjectUnloading(object? sender, EventArgs args)
             }
         }
 
-        /// <summary>
-        /// Gets a value indicating whether deleting a given set of items from the project, and optionally from disk,
-        /// would be allowed. 
-        /// Note: CanRemove can be called several times since there two types of remove operations:
-        ///   - Remove is a command that can remove project tree items from the tree/project but not from disk. 
-        ///     For that command requests deleteOptions has DeleteOptions.None flag.
-        ///   - Delete is a command that can remove project tree items from project and from disk. 
-        ///     For this command requests deleteOptions has DeleteOptions.DeleteFromStorage flag.
-        /// We can potentially support only Remove command here, since we don't remove Dependencies form disk, 
-        /// thus we return false when DeleteOptions.DeleteFromStorage is provided.
-        /// </summary>
-        /// <param name="nodes">The nodes that should be deleted.</param>
-        /// <param name="deleteOptions">
-        /// A value indicating whether the items should be deleted from disk as well as from the project file.
-        /// </param>
-        public override bool CanRemove(IImmutableSet<IProjectTree> nodes,
-                                       DeleteOptions deleteOptions = DeleteOptions.None)
+        public override string? GetPath(IProjectTree node)
         {
-            if (deleteOptions.HasFlag(DeleteOptions.DeleteFromStorage))
-            {
-                return false;
-            }
+            // Returning null from a root graft causes CPS to use the "pseudo path" for the item, which has
+            // form ">123" where the number is the item's identity. This is a short string (low memory overhead)
+            // and allows fast lookup. So in general we want to return null here unless there is a compelling
+            // requirement to use the path.
+            //
+            // Analyzer diagnostics (and possibly other consumers) require a file path on items.
+
+            return node.FilePath;
+        }
 
-            DependenciesSnapshot snapshot = _dependenciesSnapshotProvider.CurrentSnapshot;
-            if (snapshot == null)
-            {
-                return false;
-            }
+        public override IProjectTree? FindByPath(IProjectTree root, string path)
+        {
+            Assumes.True(root.Flags.Contains(ProjectTreeFlags.ProjectRoot), "Expected search to operate from project root");
```

Comment:
This is not a valid assumption - why was this added?

Reply:
As a root graft, we're only called via `PhysicalProjectTreeProvider.FindByPath` and tracing back usages of that shows lots of callers that pass `tree.Root` however I cannot definitely prove we'd never receive a node other than the root. This was added during development to help me understand when that might not be the case, because if we're given a root that's within the dependencies tree itself, we would not (currently) return results.

I'll file a follow up PR that removes the `Assumes.True` line and improves the docs as below.

Reply:
I think in most cases that would be true that you are only ever passed the root, however, I've definitely written code against providers that pass in more than the root. It's very odd in fact that CPS implementation doesn't give you the root that you provided to start with.

Reply:
Yes I wondered the same, whether CPS should filter which graft providers to call based upon the root. I think the assumption is that the implementation can interpret the path in some way that gives better complexity than O(N), as this implementation does.